src/renderer/generic.zig

```diff
@@ -0,0 +1,2873 @@
+                critical.mouse,
+                critical.preedit,
+                critical.cursor_style,
+                &critical.color_palette,
+            );
+
+            // Notify our shaper we're done for the frame. For some shapers,
+            // such as CoreText, this triggers off-thread cleanup logic.
+            self.font_shaper.endFrame();
+
+            // Acquire the draw mutex because we're modifying state here.
+            {
+                self.draw_mutex.lock();
+                defer self.draw_mutex.unlock();
+
+                // Update our background color
+                self.uniforms.bg_color = .{
+                    critical.bg.r,
+                    critical.bg.g,
+                    critical.bg.b,
+                    @intFromFloat(@round(self.config.background_opacity * 255.0)),
+                };
+            }
+        }
+
+        /// Draw the frame to the screen.
+        ///
+        /// If `sync` is true, this will synchronously block until
+        /// the frame is finished drawing and has been presented.
+        pub fn drawFrame(
+            self: *Self,
+            sync: bool,
+        ) !void {
+            // We hold a the draw mutex to prevent changes to any
+            // data we access while we're in the middle of drawing.
+            self.draw_mutex.lock();
+            defer self.draw_mutex.unlock();
+
+            // There's probably a more elegant way to do this...
+            //
+            // This is effectively an @autoreleasepool{} block, which we need in
+            // order to ensure that autoreleased objects are properly released.
+            const pool = if (builtin.os.tag.isDarwin())
+                @import("objc").AutoreleasePool.init()
+            else
+                void;
+            defer if (builtin.os.tag.isDarwin()) pool.deinit();
+
+            // Retrieve the most up-to-date surface size from the Graphics API
+            const surface_size = try self.api.surfaceSize();
+
+            // If either of our surface dimensions is zero
+            // then drawing is absurd, so we just return.
+            if (surface_size.width == 0 or surface_size.height == 0) return;
+
+            const size_changed =
+                self.size.screen.width != surface_size.width or
+                self.size.screen.height != surface_size.height;
+
+            // Conditions under which we need to draw the frame, otherwise we
+            // don't need to since the previous frame should be identical.
+            const needs_redraw =
+                size_changed or
+                self.cells_rebuilt or
+                self.hasAnimations() or
+                sync;
+
+            if (!needs_redraw) {
+                // We still need to present the last target again, because the
+                // apprt may be swapping buffers and display an outdated frame
+                // if we don't draw something new.
+                try self.api.repeat();
```

Comment:
As an API name, I don't love this, its unclear what needs to be repeated. I looked at both implementations and the comments and wonder if `presentLastFrame` or something like that would be better here.